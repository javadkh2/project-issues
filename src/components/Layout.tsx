import styled from 'styled-components'
import { colors, size } from '../config/style'

const Layout = styled.div`
  font-family: 'Roboto';
  max-width: 800px;
  margin: 0 auto;
  background: ${colors.ultraLight};
  padding: ${size()};
`

export default Layout
