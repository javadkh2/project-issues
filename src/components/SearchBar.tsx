import React from 'react'
import styled from 'styled-components'
import colors from '../config/colors'
type Props = {
  defaultValue?: string
  className?: string
}

const Input = styled.input`
  border-radius: 4px;
  border: solid 1px ${colors.light};
  padding: 10px;
`

export default function SearchBar({
  defaultValue,
  className,
}: Props): JSX.Element {
  return <Input defaultValue={defaultValue} className={className} />
}
