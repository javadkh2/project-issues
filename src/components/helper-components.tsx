import styled from 'styled-components'
import { colors, size } from '../config/style'

export const Button = styled.button`
  background: ${({ secondary }) =>
    secondary ? colors.secondary : colors.primary};
  ${({ disabled }) => disabled && 'opacity: 0.5; cursor: not-allowed;'}
  color: ${colors.white};
  border: none;
  padding: ${size()};
`

export const Error = styled.p`
  color: ${colors.primary};
  margin: 0;
  flex: 1;
  text-align: ${({ align = 'right' }) => align};
  em {
    font-style: normal;
    font-weight: bold;
  }
`

export const Part = styled.div`
  margin-top: 10px;
`

export const Anchor = styled.a`
  color: ${colors.secondary};
  text-decoration: underline;
`
