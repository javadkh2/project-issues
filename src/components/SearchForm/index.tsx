import React, { useState } from 'react'
import { Filter } from '../../type'
import {
  FilterWrapper,
  Form,
  Input,
  RadioButton,
  SearchBarWrapper,
  Title,
} from './helper-cmponents'

import { Button, Error } from '../helper-components'

export type Props = {
  defaultValue?: string
  className?: string
  defaultFilter?: Filter
  onSubmit: (value: string, filter: Filter) => void
}

function validateValue(value: string) {
  const parts = value.split('/')
  return parts.length === 2 && parts.filter(Boolean).length === 2
}

export default function SearchForm({
  defaultValue,
  className,
  defaultFilter = 'all',
  onSubmit,
}: Props): JSX.Element {
  const [filter, setFilter] = useState(defaultFilter)
  const [value, setValue] = useState(defaultValue)
  const [isValid, setIsValid] = useState(true)

  const setFilterValue = (e) => {
    setFilter(e.target.value)
    submit(value, e.target.value)
  }
  const isFilter = (f) => f === filter

  function submit(value, filter) {
    const valid = validateValue(value)
    setIsValid(valid)
    if (valid) {
      onSubmit(value, filter)
    }
  }

  return (
    <Form
      data-testid="search-form"
      onSubmit={(e) => {
        e.preventDefault()
        submit(value, filter)
      }}
    >
      <Title>Enter a repository name</Title>
      <SearchBarWrapper>
        <Input
          data-testid="search-input"
          invalid={!isValid}
          type="search"
          name="search"
          value={value}
          onChange={(e) => {
            setValue(e.target.value)
            setIsValid(true)
          }}
          className={className}
        />
        <Button type="submit">Search</Button>
      </SearchBarWrapper>
      <FilterWrapper>
        <RadioButton
          text="All"
          value="all"
          name="filter"
          onChange={setFilterValue}
          checked={isFilter('all')}
        />
        <RadioButton
          text="Open"
          value="open"
          name="filter"
          onChange={setFilterValue}
          checked={isFilter('open')}
        />
        <RadioButton
          text="Closed"
          value="closed"
          name="filter"
          onChange={setFilterValue}
          checked={isFilter('closed')}
        />
        {!isValid && (
          <Error>
            The repository should be in <em>owner/repository</em> format
          </Error>
        )}
      </FilterWrapper>
    </Form>
  )
}
