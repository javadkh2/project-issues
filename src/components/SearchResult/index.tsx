import Link from 'next/link'
import { Anchor, Button } from '../helper-components'
import {
  IssuesWrapper,
  IssueWrapper,
  Pagination,
  ResultTitle,
  ResultWrapper,
} from './helper-components'

type Props = {
  issues: any[]
  page: number
  onPaginate: (page: number) => void
  owner: string
  repository: string
}

export default function SearchResult({
  issues,
  page,
  onPaginate,
  owner,
  repository,
}: Props): JSX.Element {
  return (
    <ResultWrapper>
      <ResultTitle>Result</ResultTitle>
      <IssuesWrapper>
        {issues.map(({ title, number }) => (
          <IssueWrapper key={number}>
            <Link href={`/${owner}/${repository}/${number}`}>
              <Anchor href={`/${owner}/${repository}/${number}`}>
                {title}
              </Anchor>
            </Link>
          </IssueWrapper>
        ))}
      </IssuesWrapper>
      {(page != 1 || issues.length == 10) && (
        <Pagination>
          <Button
            secondary
            disabled={page == 1}
            onClick={() => {
              onPaginate(page - 1)
            }}
          >
            Prev
          </Button>
          <span>(page: {page})</span>
          <Button
            secondary
            disabled={issues.length < 10}
            onClick={() => {
              onPaginate(page + 1)
            }}
          >
            Next
          </Button>
        </Pagination>
      )}
    </ResultWrapper>
  )
}
