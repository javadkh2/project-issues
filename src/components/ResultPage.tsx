import Head from 'next/head'
import { useRouter } from 'next/router'
import Layout from './Layout'
import SearchForm from './SearchForm'

type Props = {
  filter?: 'all' | 'open' | 'closed'
  owner?: string
  repository?: string
}

// I couldn't find a way to route different urls (/all,/open,/closed) to the same result page.
// so I had to create separate files for each path. and then exported this component from those files
export const ResultPage = ({
  filter,
  owner,
  repository,
}: Props): JSX.Element => {
  const router = useRouter()
  const defaultValue = owner && repository && `${owner}/${repository}`
  return (
    <Layout>
      <Head>
        <title>{filter}</title>
      </Head>

      <main>
        <SearchForm
          defaultValue={defaultValue}
          defaultFilter={filter}
          onSubmit={(value, filter) => {
            router.push(`/${value}/${filter}`)
          }}
        />
      </main>
    </Layout>
  )
}

// because we use this fn in deferent pages so I make it a bit general by partially pass some default props to it
export const getResultPageProps = (defaultProps = {}) =>
  async function getServerSideProps({
    params: { owner, repository },
  }): Promise<{ props: Props }> {
    return {
      props: { ...defaultProps, owner, repository },
    }
  }

export default ResultPage
