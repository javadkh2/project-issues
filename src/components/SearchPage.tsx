import get from 'lodash/get'
import Head from 'next/head'
import { useRouter } from 'next/router'
import { useState } from 'react'
import { getIssues } from '../services/github'
import { Filter } from '../type'
import { Error, Part, Content } from './helper-components'
import Layout from './Layout'
import SearchForm from './SearchForm'
import SearchResult from './SearchResult'

type Props = {
  filter?: Filter
  owner?: string
  repository?: string
  issues?: any[]
  error?: string
}

// I couldn't find a way to route different urls (/all,/open,/closed) to the same result page.
// so I had to create separate files for each path. and then exported this component from those files
export const SearchPage = ({
  filter,
  owner,
  repository,
  issues = [],
  error,
}: Props): JSX.Element => {
  const router = useRouter()
  const [loading, setLoading] = useState(false)
  const defaultValue = owner && repository && `${owner}/${repository}`
  const page = Number.parseInt(get(router, 'query.page', '1'))

  return (
    <Layout>
      <Head>
        <title>{filter}</title>
      </Head>

      <SearchForm
        defaultValue={defaultValue}
        defaultFilter={filter}
        onSubmit={(value, filter) => {
          setLoading(true)
          router.push(`/${value}/${filter}?page=1`)
          router.events.on('routeChangeComplete', () => setLoading(false))
        }}
      />
      <Content loading={loading}>
        {error && (
          <Part>
            <Error align="left">
              Error in fetching date (status-code: {error})
            </Error>
          </Part>
        )}

        {issues && (
          <SearchResult
            owner={owner}
            repository={repository}
            issues={issues}
            page={page}
            onPaginate={(newPage) => {
              setLoading(true)
              router.push(`/${owner}/${repository}/${filter}?page=${newPage}`)
              router.events.on('routeChangeComplete', () => setLoading(false))
            }}
          />
        )}
      </Content>
    </Layout>
  )
}

// because we use this fn in deferent pages so I make it a bit general by partially apply some default props to it
export const getResultPageProps = (defaultProps: any = {}) =>
  async function getServerSideProps(context): Promise<{ props: Props }> {
    const owner = get(context, 'params.owner', '')
    const repository = get(context, 'params.repository', '')
    const page = Number.parseInt(get(context, 'query.page', '1'))
    try {
      const issues =
        owner && repository
          ? await getIssues(owner, repository, defaultProps.filter, page)
          : null
      return {
        props: { ...defaultProps, owner, repository, issues },
      }
    } catch (e) {
      return {
        props: { ...defaultProps, owner, repository, error: e.message },
      }
    }
  }

export default SearchPage
