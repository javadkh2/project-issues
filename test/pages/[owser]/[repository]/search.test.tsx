import React from 'react'
import Home from '../../../../src/pages'
import { render } from '../../../testUtils'
import { Props } from '../../../../src/components/SearchPage'

describe('Search page', () => {
  it('matches snapshot', () => {
    const props: Props = {
      filter: 'all',
      owner: 'testOwner',
      repository: 'testRepo',
      issues: [],
    }
    const { asFragment } = render(<Home {...props} />, {})
    expect(asFragment()).toMatchSnapshot()
  })
})
