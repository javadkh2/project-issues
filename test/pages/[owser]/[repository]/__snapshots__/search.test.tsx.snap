// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ hWYzqR"
  >
    <div
      class="sc-gsTCUz sc-eCssSg JMcbo kzlSGG"
    >
      <form
        class="sc-iqHYGH MVvsM"
        data-testid="search-form"
      >
        <h2
          class="sc-jrAGrp jehxKj"
        >
          Enter a repository name
        </h2>
        <div
          class="sc-crrsfI hiXqPz"
        >
          <input
            class="sc-kEjbxe hsPXNJ"
            data-testid="search-input"
            name="search"
            type="search"
            value="testOwner/testRepo"
          />
          <button
            class="sc-dlfnbm bzkdEC"
            type="submit"
          >
            Search
          </button>
        </div>
        <div
          class="sc-dQppl gPdaty"
        >
          <label
            class="sc-bqyKva bxqoLI"
          >
            <input
              checked=""
              class="sc-kstrdz ldzzjK"
              data-testid="filter-all"
              name="filter"
              type="radio"
              value="all"
            />
            All
          </label>
          <label
            class="sc-bqyKva bxqoLI"
          >
            <input
              class="sc-kstrdz ldzzjK"
              data-testid="filter-open"
              name="filter"
              type="radio"
              value="open"
            />
            Open
          </label>
          <label
            class="sc-bqyKva bxqoLI"
          >
            <input
              class="sc-kstrdz ldzzjK"
              data-testid="filter-closed"
              name="filter"
              type="radio"
              value="closed"
            />
            Closed
          </label>
        </div>
      </form>
    </div>
    <div
      class="sc-gsTCUz JMcbo"
    >
      <div
        class="sc-iBPRYJ eyuLIv"
      >
        <div
          class="sc-idOhPF DHnoZ"
        >
          <h3
            class="sc-bkzZxe gVzoH"
          >
            Result
          </h3>
          <ul
            class="sc-fodVxV lghLof"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
